doAutoCluster method for Animal and Block

Nabarb opened the pull request:
doAutoClusters is a doMethod to perform automatic spike clustering in different units. It can be called at the block level or at the animal level: when performed on an animal it stacks together all the spikes belonging to the same channel and the performs clustering. If channels are different between blocks, the clustering is done on the most restrictive set. The user is notified of the event.

The method can be called with channels as inputs or units as input. If channels are provided the clustering is done on the specified subset of channels, If units are provided the clustering is performed on all the specified units on the  specified channels.

m053m716 committed:
Make new branch with most recent versions of **non-ported** video-scoring code from rc-proj repo.

m053m716 committed:
Make new branch with most recent versions of **non-ported** video-scoring code from rc-proj repo.

m053m716 committed:
Add utilities that are currently used by **non-ported** version of **scoreVideo**
These may be deprecated, but keeping them in to start until they are determined to be unneeded.

m053m716 committed:
Add utilities that are currently used by **non-ported** version of **scoreVideo**
These may be deprecated, but keeping them in to start until they are determined to be unneeded.

m053m716 committed:
**scoreVideo** should be a method of **Block**

m053m716 committed:
**scoreVideo** should be a method of **Block**

m053m716 committed:
Merge branch 'feature/vidScore' of https://github.com/m053m716/ePhys_packages into feature/vidScore

m053m716 committed:
ReadBlockChannelInfo optional function for handling "RC" data blocks _ChannelInfo.mat file is complete.

m053m716 committed:
Add utils.initChannelStruct and utils.desiredHeaderFields so that the extraction process can be standardized for different formats more easily by referencing these two functions from the extraction functions that are unique to different acquisition formats.
Add +workflow folder (remove +convert) and move ReadMatInfo (ad hoc for Alberto) and new functions for "conversion" and MatFile "extraction" to this folder, which is where custom function handles can be added in the future for different workflows.

m053m716 committed:
Change name of desiredHeaderFields utility to initDesiredHeaderFields so that it is grouped with other two added "init" utilities.
Add initSpikeTriggerStruct so that it will enforce a common header struct for spike parsing when integrated into TDT extraction, for example.

m053m716 committed:
Added parsing for converting old spikes format (for RC project) and two more utilities: one is a method of Block named matchProbeChannel that does exactly that. The other is utils.getChannelNum, which was previously a function inside one of the ReadMat files but should be a utility since it should be standardized for the ReadHeader functions.

m053m716 committed:
Merge branch 'feature/oldBlock_conversion' into feature/vidScore

m053m716 committed:
Bug fixes on linkStreamsField (update percent-complete properly) and readMatInfoRC ad hoc function handle (assign channel ID correctly). Moved fixNamingConvention and makeDiskFile helper functions out of tdt2Block, rhs2Block, and rhd2Block, and into their own functions in nigeLab.utils.fixNamingConvention and nigeLab.utils.makeDiskFile. Added input parameter checks to nigeLab.utils.makeDiskFile so that it doesn't just delete an existing file and then cause an error if one of the struct fields was not properly included. Updated Block.checkSpikeFile method so that when it does the conversion, it can take the input argument either as a channel index or as the char array of the channel file name and parse the channel number and probe index from that. Fixed a bug that was causing Artifact to be concatenated along the wrong dimensions.

m053m716 committed:
Add blockObj.linkVideosField, as well as corresponding +defaults.Block cell elements corresponding to new Videos fieldtype, and Public Videos property for Block class.

Nabarb committed:
Commit due to compatibility issues

I have to do this commit so that I can change the working directory. I'm having trouble developing on the remote server

Nabarb committed:
Commit due to compatibility issues

I have to do this commit so that I can change the working directory. I'm having trouble developing on the remote server

Nabarb committed:
Auto stash before merge of "feature/multiAnimals" and "origin/feature/multiAnimals"

Nabarb committed:
Auto stash before merge of "feature/multiAnimals" and "origin/feature/multiAnimals"

m053m716 committed:
Move printWarningLoop to utils per FB request

m053m716 committed:
Bugfix where parsing for OldFileNames led to error when fieldname was deblanked to an empty string. Add appropriate TAG.Videos field in defaults.Block. Fix minor field naming bug in utils.printWarningLoop and also added a line to actually delete the figure it creates so the function doesn't hang indefinitely.

m053m716 committed:
Add getFieldType method, which just returns the FieldType corresponding to a given field, as well as the total number of fields of that type across all Fields.

m053m716 committed:
Changed getFileType method so that output name is more explicit and also to give a proper error message if the input is spelled incorrectly or something.

m053m716 committed:
Add Block.getFieldTypeIndex method, as well as associated Property of Block class (ViableFieldTypes).

m053m716 committed:
Add Block.getStreamsFieldIndex to allow easier indexing/matching of Streams by Fields corresponding to signal_type.

m053m716 committed:
Add more helper functions to +utils. fread_QString is a helper function from Intan-provided Matlab extraction code that reads Qt-formatted strings. plural just puts an s or not depending on if a word should be pluralized due to a number earlier in the sentence. printLinkFieldString is for the Command Window updates that show when Block.linkToData is called.

m053m716 committed:
Fix bugs based on miscommunication with FB regarding Block.Streams structure (struct array vs array fields for each StreamType, we are going with the latter option).

m053m716 committed:
Merge branch 'feature/multiAnimals' into feature/vidScore

# Conflicts:
#	+nigeLab/+defaults/Block.m
#	+nigeLab/@Block/Block.m
#	+nigeLab/@Block/doRawExtraction.m
#	+nigeLab/@Block/initStreams.m
#	+nigeLab/@Block/linkStreamsField.m
#	+nigeLab/@Block/linkTime.m
#	+nigeLab/@Block/private/ReadRHDHeader.m

m053m716 committed:
Again, trying to keep things current during parallel development so that it is less messy to merge conflicts at the end.
Merge branch 'dev' into feature/vidScore

# Conflicts:
#	+nigeLab/+defaults/Block.m
#	+nigeLab/+utils/initChannelStruct.m
#	+nigeLab/+utils/initDesiredHeaderFields.m
#	+nigeLab/+workflow/rc2Block.m
#	+nigeLab/+workflow/readMatInfo.m
#	+nigeLab/+workflow/readMatInfoRC.m
#	+nigeLab/@Block/Block.m
#	+nigeLab/@Block/doRawExtraction.m
#	+nigeLab/@Block/init.m
#	+nigeLab/@Block/initStreams.m
#	+nigeLab/@Block/linkChannelsField.m
#	+nigeLab/@Block/private/ReadRHDHeader.m
#	+nigeLab/@Block/private/ReadRHSHeader.m
#	+nigeLab/@Block/private/ReadTDTHeader.m

m053m716 committed:
Add new classes for VideosFieldType and VidStreamsType. Parsing to link those should be working more or less now. Added a LOT to the defaults.Video configuration and comments to go along with what I was thinking there in terms of how those parameters are set up. Almost ready to start integrating with the existing blockObj.scoreVideo. -MM

m053m716 committed:
Fix minor bugs that don't allow proper update of Status for VidStreams and Videos.

m053m716 committed:
Move old "link" methods to deprecated folder.

m053m716 committed:
linkEventsField and initEvents should be working properly now.

m053m716 committed:
Add utilities- utils.getNigeLink is for returning string for fprintf command window print to link to a specific function within a class, which is useful during debugging particularly in large classes. Add makeHash- utils.makeHash returns a hash 16-digit alphanumeric string that is randomly generated which can be used to link up table RowNames, for adding to or updating metadata about scoring or whatever.

m053m716 committed:
Add new methods to Block- Block.addScoringMetadata is a method that uses the utils.makeHash to append Scoring Metadata to the Block property field Block.Scoring. Block.getEventsIndex returns the index from the Events property that corresponds to indices of a specific Field. Block.setEventData allows safe setting of Event DiskData properties, with indexing even!

m053m716 committed:
Add checks regarding issue #38 -- some meta fields hardcoded in DashBoard.

m053m716 committed:
nigeLab.Block.scoreVideo should be functioning correctly now.

m053m716 committed:
Move alignVideo to method of Block and begin porting that from old version.

m053m716 committed:
Made changes to nigeLab.utils.uidropdownbox to incorporate color scheme in nigeLab.defaults.nigelColors. Added nigeLab.utils.uiHandle class (small/simple class) to act as a handle object that stores data to be passed between ui objects. Added nigeLab.utils.uidropdownradiobox, which is a slightly more sophisticated version of nigeLab.utils.uidropdownbox that has radio buttons below the popupmenu Style control, which has its list options updated as different radio buttons are toggled.

m053m716 committed:
Bugfix for DiskData, conversion functions.

m053m716 committed:
Merge branch 'feature/nigelDialogs/dropdownmenus' into feature/vidAlign

m053m716 committed:
Couple of tweaks to uidropdownbox and uidropdownradiobox.

m053m716 committed:
Add a lot of documentation to nigeLab.libs.nigelPanel.

m053m716 committed:
Restructure nigelPanel so that logic is less confusing in constructor

m053m716 committed:
Improved integration of vidScore with nigeLab.libs.nigelPanel so that other things are more flexible for customization there in the future.

m053m716 committed:
Add a lot of documentation to alignInfo object. Fix bugs with swapping colors on strings for graphicsUpdater object.

m053m716 committed:
Update to incorporate more "standard nigeLab colors" options; also now appropriately throws a warning and returns "Black" if the requested type doesn't match any of the case statements (otherwise before it was causing errors that were hard to figure out when there was a typo).

m053m716 committed:
Update header read functions and initialization so that fields such as 'input_impedance' are not added for 'Streams' objects. Added an input configuration for nigeLab.utils.initChannelStruct where the second argument can be an existing 'channelStruct', which is then used to enter the matching "valid" fields of whatever channel struct type is specified by FieldType (first input argument). This is useful in the Intan Read Header functions, so that the Digital Streams don't need to have all the redundant information of the Channels FieldType. Added two fields to Channels and Streams default "channels struct" on initialization ('name' and 'fs') to make those properties easily accessible (kept old 'custom_channel_name' field as well for backwards compatibility). Added 'ml' (mediolateral), 'icms' (icms evoked response) and 'area' (e.g. RFA, CFA, etc) fields to 'Channels' initialization defaults since those are included for RC project and conceivably would be useful in the future to associate with 'Channels' FieldType struct.

m053m716 committed:
Cosmetic changes for uidropdownbox and uidropdownradiobox

m053m716 committed:
Add functions to utilities for converting binary stream to point-process (ts; nigeLab.utils.binaryStream2ts). Add nigeLab.utils.debouncePointProcess utility for removing grouped ts (tried to do a couple of different algorithm implementations to speed it up but the simplest loop was also timed as the fastest).

m053m716 committed:
Add a new 'do' method to Block: nigeLab.Block.doEventHeaderExtraction. This creates the "Header" event file for video scoring and alignment.

m053m716 committed:
Adjust a few initialization functions to account for slight changes in naming convention. Update "signal" class to act more conveniently on signal.belongsTo method. Add method to VidStreamsType to retrieve the desired stream, as a standardized stream type: "substreams" (updated initChannelStruct to reflect this new type). This type can match Streams from Block.Streams, but the fields are much more restricted (only 'name', 'data', and 'fs').

m053m716 committed:
Add new methods to Block: getStream, which returns a stream based on its name (e.g. if 'BeamBreak' is an element of .name field in Block.Streams, it is returned if specified using "streamName" input argument). Add guessVidStreamAlignment method to block that computes cross correlation between video stream and digital stream to make an alignment guess using utils.getR and utils.parseR functions (which are kind of slow at present).

m053m716 committed:
Complete nigeLab.Block.doEventDetection, which is a method to parse event times (e.g. transitions on 'Rising', 'Falling', or 'Any' edge for a particular stream) automatically.

m053m716 committed:
Add a bunch of error checking that throws warnings if "Event" type DiskData is accessed before it is properly initialized in one way or another (from Block.getEventData method).

m053m716 committed:
Minor bugfix for nigeLab.utils.uidropdownbox that caused an error if title_str was specified as a char array.

m053m716 committed:
Minor bugfixes for ad hoc conversion "workflow" functions for my own project

m053m716 committed:
Minor modifications to nigeLab.utils.getR and nigeLab.utils.parseR so that their inputs and output are clearer.

m053m716 committed:
**MAJOR BUG** fixed linkStreamsField method -- not sure maybe it got messed up in a previous merge? Anyhow it was behaving totally incorrectly and that was causing errors elsewhere.

m053m716 committed:
Initial working version of "guessVidStreamAlignment" although it doesn't really perform that well in practice.

m053m716 committed:
Add nigeLab.utils.applyScaleOpts and nigeLab.utils.initScaleOpts. These are utilities for scaling data from a given stream. Currently there are three automatic options - 'fixed_scale', 'normalized' (or 'unit'), and 'zscore'. Useful for things like Beam break streams that will have a constant offset

m053m716 committed:
Add '.t' as a default field for "substreams". Update getStream methods of Block and VidStreamsType to reflect new format as well as the scaling from scaleOpts utilities.

m053m716 committed:
Minor bugfix for getDropDownRadioStreams, where it wasn't parsing empty Fields options (e.g. remove AnalogIO radiobutton if there are no AnalogIO streams detected).

m053m716 committed:
Core functionality of alignInfo class and alignVideoManual methods are implemented. Now just updating cosmetic changes and fixing initialization order to make them work more smoothly.

m053m716 committed:
Fix minor bugs with initialization of alignment interface. Some cosmetic changes as well.

m053m716 committed:
Add more methods to nigelPanel for improving cosmetics and dynamically updating graphics properties on PostSet from property changes.

m053m716 committed:
Tested with different selections of input parameters. Should all be working.

m053m716 committed:
Bugfix for issue #10.
Add compatibility checks for "special" fields: 'Time', 'Probes', and 'Notes'.
Update 'NotesUI' so that 'Notes' Status can be updated dynamically on calls to blockObj.linkNotes().
Tested blockObj.Status and it no longer initializes to True by default for these Fields.

m053m716 committed:
Made some changes to try and improve compatibility with UNIX.

m053m716 committed:
Update Block.getStatus to incorporate Channel Mask if the "stage" is a Field that corresponds to "Channels" FieldType.

m053m716 committed:
Add a lot of documentation to Tank and Animal. Add 'Empty' static method to each class so that can be used during 'MultiAnimal' stuff.

m053m716 committed:
Move file to deprecated folder

m053m716 committed:
Move file to deprecated folder

m053m716 committed:
Minor changes to +workflow functions to account for change property name -- BlockPars --> PathExpr  --

m053m716 committed:
Split up parsing into a few methods of Block so that it is easier to handle cases for example when you want to parse the file hierarchy because there is no Block file saved but it has already been extracted but there isn't a ChannelInfo file, etc

m053m716 committed:
Change defaults files to include .FolderIdentifier default, which is used to put a very small "nigelFile" inside the Block or Animal or Tank folder to denote that it corresponds to that class.

m053m716 committed:
Updated methods of Block to reflect removed properties and changes to class property validation

m053m716 committed:
Update Tank and Tank methods to have similar property validation and remove unwanted properties. Did some troubleshooting on issues with loadobj and saveobj.

m053m716 committed:
Updated Animal as well to have property validation, and also fixed issues with loadobj and saveobj that were causing an infinite recursion when you tried to load it. Now it it works to load any kind of nigelObj.

m053m716 committed:
Modifications to DashBoard with a lot of documentation or splitting methods into smaller or more appropriately-named methods. Just trying to make sure that everything here works correctly with my "RC" workflow in addition to the other test cases that have been used in the past. This commit puts me at a state where DashBoard can be loaded for tankObj, and when AnimalObj is clicked it works, but when BlockObj is clicked there is an error based on Status. So essentially, this is where the actual 'doActions' modifications technically begin.

m053m716 committed:
Add {} indexing to subsref method of nigeLab.Animal. This should make it easier to reference Animal/Block combinations such as in SelectedItems from nigeLab.libs.DashBoard.

Nabarb committed:
MuLtinAnimal update

evt : readded jobComplete event that went missing for some reason
splitCompleted event added to trigger updates in the DashBoards

libs: splitmultianimalsUI this is the Gui for the multianimals interface. It takes a DashBoardObj as input during construction and communicates back and forth with it using mostly events.

utils : multiCallBackWrap is a wrapper to execute multiple functions during a callback

Block : changed all reference to ManyAnimals in MultiAnimals

reportProgress update : reportprogress now checks the last displayed characters to decide whether to update only the completed percentage or print everything back again. 
It also understands if it's ini worker environment or not, 

SplitMultiAnimals is defined both as a block and an animal method. In blocks it looks for the multiAnimal character and creates two children copy block stored in MultiAnimalsLinkedBlocks. 
In animals does the same, but with animals; 
If a TreeObj is passed as an input it applies the changes stored in the tree;

UpdatePaths now removes the old Block savefile.

Nabarb committed:
MuLtinAnimal update

evt : readded jobComplete event that went missing for some reason
splitCompleted event added to trigger updates in the DashBoards

libs: splitmultianimalsUI this is the Gui for the multianimals interface. It takes a DashBoardObj as input during construction and communicates back and forth with it using mostly events.

utils : multiCallBackWrap is a wrapper to execute multiple functions during a callback

Block : changed all reference to ManyAnimals in MultiAnimals

reportProgress update : reportprogress now checks the last displayed characters to decide whether to update only the completed percentage or print everything back again. 
It also understands if it's ini worker environment or not, 

SplitMultiAnimals is defined both as a block and an animal method. In blocks it looks for the multiAnimal character and creates two children copy block stored in MultiAnimalsLinkedBlocks. 
In animals does the same, but with animals; 
If a TreeObj is passed as an input it applies the changes stored in the tree;

UpdatePaths now removes the old Block savefile.

m053m716 committed:
Change getNumBlocks so that it handles array input for animalObj.

m053m716 committed:
Add handling of Tank{} indexing as shortcut to access child of child Blocks or child Animals.

m053m716 committed:
Changed handling for ambiguous input name call for nigelPanel.getChild method.

m053m716 committed:
Add skeleton of "nigeLab" function for top-level UI

Commit:
Merge branch 'feature/multiAnimals' into local/doActions

Commit:
Minor/Major bugfixes

DashBoard - fix typo on splitMultiAnimalsUI
Tank - Fix messed up syntax from merge. Change a bunch of bugs in algorithm for CheckAnimalsForClones.
Block - Remove references to "Parent" Animal property.
Animal - Fix syntax errors from merge. Fix bugs in algorithm for CheckBlocksForClones.

Commit:
DashBoard: apply MM stashed changes

Commit:
Clean up Listener syntax

Many of the listeners were added without a reference to the listener object. I keep the listener object as properties of Tank, Animal, or Block and then ensure that during saveObj and delete that those property values are either transiently unassigned or deleted appropriately. In addition, during call to Tank.assignNULL or Animal.assignNULL, the 'Listener' property of the corresponding child object is the actual listener handle for that event (to avoid creating an array of event listeners with messy indexing and an ever-expanding array of deleted listener handles).

Commit:
Minor bugfixes

Tank: Tank properties were not SetObservable.
Fixed bad reference to tankObj.Animals.Blocks (should just be tankObj.Animals). Changed addBlock to addChildBlock to prevent accidental referencing of Simulink method.

Header parsing: Fix errors in ReadRHSHeader and ReadRHDHeader where they were initializing DigIO and AnalogIO to 'Channels' rather than 'Streams'

Animal: 
init- Change "addBlock" variable to "addThisBlock" to avoid overlap with Simulink method. Fix reference to addBlock method (to addChildBlock).
Generally, change references from addBlock to addChildBlock.

DashBoard: Reconcile more merge changes

Commit:
Move large methods to separate files

Block  -- Moved subsref to its own file. Ideally we will eventually turn this into a .p file or something like that so it is really hard to mess with because the bulk of the problems come from this thing.

Animal  --  Move subsref to its own file for same reason as Block (although this one is less troublesome at least). Move addChildBlock to its own file as well, since that was getting large.

Tank  --  Move subsref and addAnimal to own files.

Commit:
Cleanup subsref

Major changes to nigeLab.Block subsref to make the shortcuts cleaner and reduce chances of returning an unexpected result.

Commit:
(no branch): (no branch): Auto stash before revert of "Add link to corresponding extraction methods"

I broke my git and I'm still scared curse you Kraken

Commit:
More of me breaking my Github

Why did I try to revert that one stupid commit

Commit:
Redo changes from GitFail: part 1

Redo IsEmpty stuff for Block, Animal, Tank.

Commit:
More GitHub failing

I lost even more changes than I thought?

Commit:
GitFail cleanup: Part 2

Fix callbacks to listeners that were previously working then stopped working because I can't revert a tiny commit without destroying the whole Git Flow.

Commit:
GitFail cleanup: Part 3

I think finally back to where I was before Kraken got me

Commit:
GitFail cleanup: Part 4

Oh my god I lost the whole day because of one stupid button click.

Commit:
GitFail cleanup: Part 5

Almost back to where I was. Fixing methods in nigeLab.libs.DashBoard

Nabarb committed:
doAutoCluster update and fixes

update doAutoCluster method to make it actual.
Added a defaults ad-hoc folder to host custom clustering methods defaults. They can be found in pars.AutoClustering.METHOD. At this point KMEANS and SPC are implemented.

The method doAutoClustering calls submethods depending on what defined in defualts.

Some other small fixes.

Nabarb committed:
doAutoCluster update and fixes

update doAutoCluster method to make it actual.
Added a defaults ad-hoc folder to host custom clustering methods defaults. They can be found in pars.AutoClustering.METHOD. At this point KMEANS and SPC are implemented.

The method doAutoClustering calls submethods depending on what defined in defualts.

Some other small fixes.

Commit:
Add nigelButton class

Utility class for creating the buttons with rounded corners.

Commit:
Minor adjustments associated with nigelButton

Update parsing of inputs to nigelColors to include 'button' and 'onbutton' cases.

Added example usage to nigeLab.libs.nigelButton.

Minor formatting things for tabs on nigelPanel

Nabarb committed:
Animal doAutoClustering update

doAutoClustering method now works at the animal level.

Nabarb committed:
Animal doAutoClustering update

doAutoClustering method now works at the animal level.

Commit:
Saving progress

Tank - fixed subsref error with shortcut subscripting via {}. Fixed `isempty` overload.

Block - Add "ProgressChanged" and "MethodCanceled" events, although "MethodCanceled" may go away. Re-added reload method. Fixed bug in `initStreams` causing bad initialization. Added some progress reporting for Serial progress reports on `intan2Block`. Added documentation to `reportProgress` and also moved a copy to `nigeLab.utils` for possible future use. Made `reportProgress` optionally accept an additional 4th argument to specify "notification_mode"; if set as 'toEvent' this will notify the nigelProgressObj (progress bar in DashBoard) and skip updating the command window. Removed 'Notifications' from list of props to skip on updateParams.

Animal - fixed same issue with `isempty` overloaded method.

Defaults - added argument parsing to `Notifications`.

evt - Modified properties of `jobCompletedEventData` and added `progressChangedEventData`, which is used in `reportProgress` method of `Block`.

DashBoard - Extensive changes to make it work with more listener events. Bars should go into correct locations now when job is finished running, although the bar will "disappear" when the process stops. 

nigelButton - Fix class destructor

nigelProgress - This is where most of the listener syntax changes were made.

remoteMonitor - Made changes to reflect those made in nigelProgress

utils - Added comments to `checkforworker`, fixed an error in `fixNamingConvention` that was causing empty headers to throw errors, added comments to `getLastDispText`, fix bug in variable name of `getNormPixelDim`, fixed empty assignment bug in `initChannelStruct`, and added copy of `reportProgress` without the `notify` part.

Commit:
Current status: multiAnimals not working, DashBoard populates

Committing and pushing all local changes to remote for further dev by FB.

Commit:
Merge branch 'local/doActions' into feature/doActions

Commit:
Merge remote-tracking branch 'origin/feature/multiAnimals' into feature/multiAnimals

Commit:
Merge branch 'feature/doActions' into feature/multiAnimals

Trying to collect everything to push for FB.

Commit:
Attempting to fix multiAnimals interface

Added "border highlighting" to buttons on button down click, which will "un-highlight" prior to execution of the actual button callback. Unsuccessfully trying to debug the multianimals thing.

Commit:
Current status: multiAnimals not working

Right now it cycles through and deletes all the child blocks and the child "multi-animal" as well and doesn't reload anything. The "tree" interfaces do not populate. I think it's probably something in either Block or Animal but not sure what.

nigelColors  --  Docs formatting

DashBoard  --  Added SelectionIndex SetObservable property that works in combination with new features on nigelButton to prevent "Split" option from enabling on inappropriate nodes.

nigelButton  --  Added 'Enable' SetObservable property and changed the way the ButtonDownFcn works from being on a listener that updates the functions to having it be split into two properties: 'Fcn' and 'Fcn_Args' which are then called from a static 'ButtonDownFcn' assigned to the rectangle and text graphics. Added a 'highlight' edge to indicate that the button has been clicked, which stops glowing when button is released.

splitMultiAnimalsUI  -- Went through and commented on a bunch of stuff trying to understand how it works and add some basic documentation to it. Added property validation to different properties.

Did the same thing as for the libs class for both associated methods of nigeLab.Block and nigeLab.Animal as well.

Commit:
New outline for cleaning up splitMultiAnimalsUI

Use restricted properties of nigeLab.libs.DashBoard to point to the "animals and blocks to split" so that they can be "cleaned up" after the splitting procedure is finished. These properties can only be accessed by nigeLab.libs.DashBoard and nigeLab.libs.splitMultiAnimalsUI (.A_split and .B_split).

Commit:
Final changes before back to FB for splitMultiAnimals

Fixed bug with Tank subsref via {} that caused Animals to return as column array instead of row array.

Added '.toSplit' and '.toAdd' property struct arrays to nigeLab.libs.DashBoard, which can be used to keep track of what should be added or removed from from the DashBoard Tree interface.

Nabarb committed:
Merge remote-tracking branch 'origin/feature/doAutoCluster' into feature/doAutoCluster

Nabarb committed:
bugfixes

Small bugfixes

Nabarb committed:
Autocluster update part 2

Corrected some calls to the wrong defaults

doAutocluster is formatted better and in working order both for animal and block

getSort now copies all Clust files into sort when Sort files are not there or not initialized

Nabarb committed:
Bugfixes

Small bugfixes 
Block DynamicVarExp needs to be init as cell
doMethods, corrected for the new Pars syntax. Other corrections will probably bee needed in the future.
Corrected also for notifyUser/reportProgress

notifyUser moved to deprecated

linkToData added case where no files are present in the the target folder.

parseNamingMetadata small fix

Nabarb committed:
Merge branch 'dev' into feature/doAutoCluster

Nabarb committed:
Bugfix

Small bugfixes

Nabarb committed:
Merge branch 'dev' into feature/doAutoCluster

Nabarb committed:
doAutoClustering update final

doAutoClustering is now working on animal and block

when performed on an animal it stacks together all the spikes belonging to the same channel, before clustering. If channels are different between blocks, the clustering is done on the more restrictive set. The user is notified of the event.

doSD does not initialize Clusters and Sorted anymore

m053m716 commented:
My inclination is to add it to docs that 'dot' indexing after '{}' shortcut is not supported. It just splits the code up to two lines: first grabbing the selected Block, then performing the selected method or referencing the desired Block property.

Commit:
Add improved parsing for initialization

Primary changes are in nigeLab.defaults.Block, where there are a few added fields that make the parsing more customizable (e.g. ability to split up the 'AnimalID' into multiple separate variables: delim can be increased to {'_', '-'} in a new 'VarExprDelimiter' field without breaking backwards compatibility (hopefully). This lets user customize the way that 'AnimalID' and 'RecID', which are now "special" fields of 'meta' should be parsed, such that they will always have values (making "downstream" stuff.
I realize these should go on a different branch, but with those changes, this should be accepted and merged into dev anyhow (pending FB approval).
